libbpf-tools/map_helpers.c

```diff
@@ -0,0 +1,105 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2020 Anton Protopopov
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+
+#include "map_helpers.h"
+
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static bool batch_map_ops = true; /* hope for the best */
+
+static int
+dump_hash_iter(int map_fd, void *keys, __u32 key_size,
+	       void *values, __u32 value_size, __u32 *count,
+	       void *invalid_key)
+{
+	__u8 key[key_size], next_key[key_size];
+	__u32 n = 0;
+	int err;
+
+	/* First get keys */
+	__builtin_memcpy(key, invalid_key, key_size);
+	while (n < *count) {
+		err = bpf_map_get_next_key(map_fd, key, next_key);
+		if (err && errno != ENOENT) {
+			return -1;
+		} else if (err) {
+			break;
+		}
+		__builtin_memcpy(key, next_key, key_size);
+		__builtin_memcpy(keys + key_size*(n++), next_key, key_size);
```

Comment:
nit: it's correct, but it's hard to notice that increment here, it's a good idea to not do such increments in complicated expressions like this

Reply:
yeah, moved to the next line